### PR TITLE
Add a parameter to allow application to reset time after restart

### LINKED ISSUE #11223: Have a parameter to reset time after restart
## Rationale

We have a transient master and a transient subapp, both restarted from a steady-state solution with a steady-state master and a transient subapp (they are doing picard iteration, subapp reach steady-state in each picard iteration). Now the subapp keeps the time reaching steady-state from the steady-state calculations on initial, which is meaningless. Because the transient starts from time 0, transient subapp will skip evaluations between 0 and the restarted time, which is not what we want. We can change the start time to a time bigger than the restarted time, but that is not very ideal.

## Description

We can add a parameter `reset_time` like `reset_dt` in `Transient` to reset the time to the app starting time after restart. I thought about to add this parameter in `TransientMultiApp`, but that will make resetting time completely after `Transient::init`, which might introduce subtle bugs because `Transient::init` calls time-integrator init. Actually, the more I think, this parameter possibly needs to be in `FEProblemBase` and will be used inside `initialSetup`. Comments?

## Impact
New capability.

### PR DESCRIPTION
**Pull request by YaqiWang**
Close #11223. If you want a test, let me know. Also I am not sure if this is the proper design, because I do not know what is the difference between `restartFromFile` and `restartRestartableData` with a parameter `skip_additional_restart_data`. Please note that this is not implemented the same as `reset_dt` in `Transient`.

### CODE REVIEW

File: framework/src/problems/FEProblemBase.C [hunk 1]
```diff
@@ -665,6 +670,8 @@ FEProblemBase::initialSetup()
     // for some of the variables which should override what's coming from the restart file
     if (!_app.isRecovering())
     {
+      if (getParam<bool>("reset_time_after_restart"))
+        _time = _time_old = _app.getStartTime();
```
Comment (friedmud) [inline framework/src/problems/FEProblemBase.C]: This seems like a decent plan.

File: framework/src/problems/FEProblemBase.C [hunk 2]
```diff
@@ -139,6 +139,11 @@ validParams<FEProblemBase>()
                         false,
                         "True to skip additional data in equation system for restart. It is useful "
                         "for starting a transient calculation with a steady-state solution");
+  params.addParam<bool>(
+      "reset_time_after_restart",
```
Comment (friedmud) [inline framework/src/problems/FEProblemBase.C]: It makes a bit more sense to me for this to be in the Executioner block.

**Review by friedmud — CHANGES_REQUESTED** [review]
Should the new parameter be in the `Executioner` block?

File: framework/src/problems/FEProblemBase.C (hunk 2 above)
Comment (YaqiWang) [reply framework/src/problems/FEProblemBase.C]: I thought about this. If we did not reset time right after it is restarted, we could potentially create inconsistency on the initial evaluations, etc. I initially thought I can make this parameter in `TransientMultiApp`, but that makes things even more delayed and could cause subtle bugs.

**Comment by friedmud** [comment]
_when_ doesn’t have anything to do with _where_.

We can make an Action happen whenever we want. The question is: where is
the right place to put this parameter?

To me: it’s in the same place we are setting the initial time, etc.

**Comment by YaqiWang** [comment]
That is impossible unless we let `FEProblemBase::initialSetup` accept an argument. From the place where time is restarted to the end of the function, there are tons of stuff going on, which might depend on the time... The starting time is an arguable parameter, since MooseApp actually own it even `Transient` accepts it from user's input. This is the minimum intrusive solution.

**Comment by permcody** [comment]
> That is impossible unless we let FEProblemBase::initialSetup accept an argument. 

No, it's not. Take a look at https://github.com/idaholab/moose/blob/devel/framework/src/actions/CreateProblemAction.C#L88-L93

We build the problem and set a bunch of parameters on it already.

**Comment by friedmud** [comment]
Add a new action that acts on “/Executioner” that takes this parameter
(defaulted to false) and sets it on the Problem. Stick it into the task
pipeline as early as you want.

**Comment by permcody** [comment]
Closing per the 30-day limit. Please feel free to reopen when you have a chance to address comments.

**Pull request closed without merging**